### PR TITLE
Pure uhd develop

### PR DESCRIPTION
**Pull request by Shadowcaster77**
Several things have been added and changed:
1. PURE UHD version of Radio, ClientRadioSet and BaseRadioSet
2. receiver.h and receiver.cc update to make changes compatible
3. several new configuration files in special_conf
4. slight update in post python processing in hdf5_lib.py

### CODE REVIEW

File: CC/Sounder/ClientRadioSetUHD.cc [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+#include "include/RadioUHD.h"
+#include "include/comms-lib.h"
+#include "include/logger.h"
+#include "include/macros.h"
+#include "include/utils.h"
+
+using json = nlohmann::json;
+
+//static void initAGC(uhd::usrp::multi_usrp::sptr* dev, Config* cfg);
```
Comment (asedlmayr) [inline CC/Sounder/ClientRadioSetUHD.cc:19]: Remove dead code

File: CC/Sounder/ClientRadioSetUHD.cc [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+#else
+    init(context);
+#endif
+  }
+  // Wait for init
+  while (thread_count.load() > 0) {
+  }
+  // commented out for now, will add this when future doing Multi USRP
+  // Strip out broken radios.
+  // Update for UHD multi USRP
```
Comment (asedlmayr) [inline CC/Sounder/ClientRadioSetUHD.cc:53]: Remove or put in an if statement that compiles out

File: CC/Sounder/ClientRadioSetUHD.cc [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    MLPD_WARN("Unknown exception\n");
+    if (radios != nullptr) {
+      delete radios;
+      radios = nullptr;
+    }
+    throw;
+  }
+  if (has_runtime_error == false) {
+    //        auto dev = radios.at(i)->dev;
+    //        SoapySDR::Kwargs info = dev->getHardwareInfo();
```
Comment (asedlmayr) [inline CC/Sounder/ClientRadioSetUHD.cc:184]: remove dead code

File: CC/Sounder/ClientRadioSetUHD.cc [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+      std::cout << "check ch: " << ch << std::endl;
+      auto new_ch = _cfg->cl_channel();
+      double rxgain = _cfg->cl_rxgain_vec().at(ch).at(
+          i);  // w/CBRS 3.6GHz [0:105], 2.5GHZ [0:108]
+      double txgain = _cfg->cl_txgain_vec().at(ch).at(
+          i);  // w/CBRS 3.6GHz [0:105], 2.5GHZ [0:105]
+      radios->dev_init(_cfg, ch, rxgain, txgain);
+    }
+
+    // Init AGC only for Iris device
```
Comment (asedlmayr) [inline CC/Sounder/ClientRadioSetUHD.cc:195]: remove

File: CC/Sounder/ClientRadioSetUHD.cc [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+    frameTime = SoapySDR::timeNsToTicks(frameTimeNs, _cfg->rate());
+
+#if DEBUG_RADIO
+    if (frameTimeNs < 2e9)
+      std::cout << "client " << radio_id << " received " << ret << " at "
+                << frameTimeNs << std::endl;
+#endif
+    //        double a = radios->dev->get_time_now().get_real_secs();
+    //        std::cout << "time: " << a <<std::endl;
```
Comment (asedlmayr) [inline CC/Sounder/ClientRadioSetUHD.cc:237]: remove

File: CC/Sounder/RadioUHD.cc [hunk 6]
```diff
@@ -0,0 +1,244 @@
+#include "include/RadioUHD.h"
+
+#include "include/comms-lib.h"
+#include "include/logger.h"
+#include "include/macros.h"
+
+#ifdef UHD_HAS_MSG_HPP
+#include <uhd/utils/msg.hpp>
```
Comment (asedlmayr) [inline CC/Sounder/RadioUHD.cc:8]: Change non system includes to use ""  instead of <>

File: CC/Sounder/RadioUHD.cc [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+  MLPD_TRACE("Number of reads needed to drain: %d\n", i);
+}
+
+RadioUHD::RadioUHD(const std::map<std::string, std::string>& args,
+                   const char uhdFmt[], const std::vector<size_t>& channels) {
+  //    dev = SoapySDR::Device::make(args);
+  dev = uhd::usrp::multi_usrp::make(args);
+  //    uhd::stream_args_t stream_args("fc32");
+  if (dev == NULL) throw std::invalid_argument("error making UHD:Device\n");
+  //    for (auto ch : channels) {
```
Comment (asedlmayr) [inline CC/Sounder/RadioUHD.cc:82]: remove

File: CC/Sounder/include/BaseRadioSetUHD.h [hunk 8]
```diff
@@ -0,0 +1,59 @@
+#include <SoapySDR/Formats.hpp>
```
Comment (asedlmayr) [inline CC/Sounder/include/BaseRadioSetUHD.h:1]: Change non system includes to use ""  instead of <>

File: CC/Sounder/include/BaseRadioSetUHD.h [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+#include <SoapySDR/Time.hpp>
+#include <uhd/usrp/multi_usrp.hpp>
+
+#include "config.h"
+#include "nlohmann/json.hpp"
+
+#ifndef BASE_RADIO_SET_UHD_H
+#define BASE_RADIO_SET_UHD_H
+
+class RadioUHD;
```
Comment (asedlmayr) [inline CC/Sounder/include/BaseRadioSetUHD.h:11]: Remove forward declarations

File: CC/Sounder/include/ClientRadioSetUHD.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+#include <SoapySDR/Formats.hpp>
+#include <SoapySDR/Time.hpp>
+
+#include "config.h"
+#include "nlohmann/json.hpp"
+
+#ifndef CLIENT_RADIO_SET_UHD_H
+#define CLIENT_RADIO_SET_UHD_H
+
+#pragma once
```
Comment (asedlmayr) [inline CC/Sounder/include/ClientRadioSetUHD.h:11]: remove

File: CC/Sounder/include/RadioUHD.h [hunk 11]
```diff
@@ -0,0 +1,46 @@
+/*
+ Copyright (c) 2018-2019, Rice University
+ RENEW OPEN SOURCE LICENSE: http://renew-wireless.org/license
+
+*/
+
+#include <SoapySDR/Device.hpp>
```
Comment (asedlmayr) [inline CC/Sounder/include/RadioUHD.h:7]: Change non system includes to use ""  instead of <>

File: CC/Sounder/include/RadioUHD.h [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+class RadioUHD {
+ private:
+  uhd::usrp::multi_usrp::sptr dev;
+  uhd::rx_streamer::sptr rxs;
+  uhd::tx_streamer::sptr txs;
+
+  void reset_DATA_clk_domain(void);
+  void dev_init(Config* _cfg, int ch, double rxgain, double txgain);
+  friend class ClientRadioSetUHD;
```
Comment (asedlmayr) [inline CC/Sounder/include/RadioUHD.h:24]: It would be nice to get rid of this "friend class stuff"  but since it is in the current code.  We can leave it here for now

File: CC/Sounder/include/receiver.h [hunk 13]
```diff
@@ -85,6 +87,9 @@ class Receiver {
   ClientRadioSet* clientRadioSet_;
   BaseRadioSet* base_radio_set_;
 
+  ClientRadioSetUHD* clientRadioSetUHD_;
```
Comment (asedlmayr) [inline CC/Sounder/include/receiver.h:90]: This over complicates the receiver class

**Review by asedlmayr — COMMENTED** [review]
Remove dead code.  
Clean up header files a little more (no forward declarations / #pragma once / and "" for non system includes)

File: CC/Sounder/include/BaseRadioSetUHD.h (hunk 9 above)
Comment (Shadowcaster77) [reply CC/Sounder/include/BaseRadioSetUHD.h:11]: the use of the forward declarations is what the current Sounder has for the Base/ClientRadioSet.h. 
Do you recommend replacing this with #include "RadioUHD.h" or keep consistent with what currently had in the develop branch

File: CC/Sounder/ClientRadioSetUHD.cc [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+*/
+#include "include/ClientRadioSetUHD.h"
+
+#include "include/RadioUHD.h"
+#include "include/comms-lib.h"
+#include "include/logger.h"
+#include "include/macros.h"
+#include "include/utils.h"
+
+static void freeRadios(RadioUHD*& radios) { delete radios; }
```
Comment (asedlmayr) [inline CC/Sounder/ClientRadioSetUHD.cc:17]: *& ... maybe there is a better way then passing a reference to a ptr for this functions

File: CC/Sounder/include/ClientRadioSetUHD.h [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+  struct ClientRadioContext {
+    ClientRadioSetUHD* crs;
+    std::atomic_ulong* thread_count;
+    size_t tid;
+  };
+  void init(ClientRadioContext* context);
+  static void* init_launch(void* in_context);
+
+  Config* _cfg;
+  RadioUHD* radios;
```
Comment (asedlmayr) [inline CC/Sounder/include/ClientRadioSetUHD.h:40]: called radios, but it is a single ptr.   (Radio)
Add an _ to the end of all member variables.

File: CC/Sounder/include/Radio.h [hunk 16]
```diff
@@ -1,25 +1,23 @@
...
 */
+#ifndef RADIO_H_
+#define RADIO_H_
 
-#include <SoapySDR/Device.hpp>
-#include <SoapySDR/Time.hpp>
+#include <cstdlib>
+#include <vector>
 
+#include "SoapySDR/Device.hpp"
```
Comment (asedlmayr) [inline CC/Sounder/include/Radio.h:13]: Do you really need to include all these soapy headers?

File: CC/Sounder/include/BaseRadioSetUHD.h [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+ * @file BaseRadioSetUHD.h
+ * @brief Declaration file for the BaseRadioSetUHD class.
+ */
+
+#ifndef BASE_RADIO_SET_UHD_H_
+#define BASE_RADIO_SET_UHD_H_
+
+#include "RadioUHD.h"
+#include "SoapySDR/Formats.hpp"
+#include "SoapySDR/Time.hpp"
```
Comment (asedlmayr) [inline CC/Sounder/include/BaseRadioSetUHD.h:11]: Do you need to include the soapy headers here?  Add necessary system headers for size_t, void, std::atomic.

Comment (asedlmayr) [reply CC/Sounder/include/BaseRadioSetUHD.h:11]: Shouldn't need json included here

File: CC/Sounder/include/BaseRadioSetUHD.h [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+  };
+  void init(BaseRadioContext* context);
+  void configure(BaseRadioContext* context);
+
+  static void* init_launch(void* in_context);
+  static void* configure_launch(void* in_context);
+
+  void radioTriggerUHD(void);
+  void sync_delays(size_t cellIdx);
+  //    SoapySDR::Device* baseRadio(size_t cellId);
```
Comment (asedlmayr) [inline CC/Sounder/include/BaseRadioSetUHD.h:48]: remove dead code

File: CC/Sounder/include/BaseRadioSetUHD.h [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+  uhd::usrp::multi_usrp::sptr baseRadio(size_t cellId);
+  int syncTimeOffsetUHD(bool, bool);
+  void dciqCalibrationProcUHD(size_t){};
+  void readSensors(void);
+
+  Config* _cfg;
+  //    std::vector<SoapySDR::Device*> hubs;
+  uhd::usrp::multi_usrp::sptr hubs;
+  //    std::vector<std::vector<Radio*>> bsRadios; // [cell, iris]
```
Comment (asedlmayr) [inline CC/Sounder/include/BaseRadioSetUHD.h:57]: remove dead code

File: CC/Sounder/include/RadioUHD.h [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+  * RENEW OPEN SOURCE LICENSE: http://renew-wireless.org/license
+*/
+#ifndef RADIOUHD_H_
+#define RADIOUHD_H_
+
+#include <cstdlib>
+#include <vector>
+
+#include "SoapySDR/Device.hpp"
+#include "SoapySDR/Errors.hpp"
```
Comment (asedlmayr) [inline CC/Sounder/include/RadioUHD.h:14]: Soapy includes necessary?

File: CC/Sounder/receiver.cc [hunk 21]
```diff
@@ -401,6 +418,7 @@ void Receiver::loopRecv(int tid, int core_id, SampleBuffer* rx_buffer) {
       size_t radio_id = it - config_->n_bs_sdrs_agg().at(cell);
       bs_sync_ret = -1;
       while (bs_sync_ret < 0) {
+        std::cout << "bs_sync_ret is being executed" << std::endl;
```
Comment (asedlmayr) [inline CC/Sounder/receiver.cc:421]: remove debug print

File: CC/Sounder/receiver.cc [hunk 22]
```diff
@@ -448,8 +467,10 @@ void Receiver::loopRecv(int tid, int core_id, SampleBuffer* rx_buffer) {
       // Set buffer status(es) to full; fail if full already
       for (size_t ch = 0; ch < num_packets; ++ch) {
         int bit = 1 << (cursor + ch) % sizeof(std::atomic_int);
+        //        std::cout<<"bit at 1 is " << bit << std::endl;
```
Comment (asedlmayr) [inline CC/Sounder/receiver.cc:470]: remove dead code

File: CC/Sounder/receiver.cc [hunk 23]
```diff
@@ -448,8 +467,10 @@ void Receiver::loopRecv(int tid, int core_id, SampleBuffer* rx_buffer) {
...
         int offs = (cursor + ch) / sizeof(std::atomic_int);
         int old = std::atomic_fetch_or(&pkt_buf_inuse[offs], bit);  // now full
+        //        std::cout<<"old at 1 is " << old << std::endl;
```
Comment (asedlmayr) [inline CC/Sounder/receiver.cc:473]: remove dead code

File: CC/Sounder/receiver.cc [hunk 24]
```diff
@@ -497,16 +520,32 @@ void Receiver::loopRecv(int tid, int core_id, SampleBuffer* rx_buffer) {
...
-                               rxTimeBs + txTimeDelta);
+                               rxTimeBs + txTimeDelta_);
           }  // end if config_->dul_data_slot_present()
         }
         if (!config_->isPilot(frame_id, slot_id) &&
-            !config_->isUlData(frame_id, slot_id))
+            !config_->isUlData(frame_id, slot_id)) {
+          for (size_t ch = 0; ch < num_packets; ++ch) {
+            const int bit = 1 << (cursor + ch) % sizeof(std::atomic_int);
+            //                std::cout<<"bit at 2 is " << bit << std::endl;
```
Comment (asedlmayr) [inline CC/Sounder/receiver.cc:534]: remove dead code

File: CC/Sounder/receiver.cc [hunk 25]
```diff
@@ -497,16 +520,32 @@ void Receiver::loopRecv(int tid, int core_id, SampleBuffer* rx_buffer) {
...
+            const int offs = (cursor + ch) / sizeof(std::atomic_int);
+            const int old =
+                std::atomic_fetch_and(&pkt_buf_inuse[offs], ~bit);  // now full
+            //                std::cout<<"old at 2 is " << old << std::endl;
```
Comment (asedlmayr) [inline CC/Sounder/receiver.cc:538]: remove dead code

File: CC/Sounder/receiver.cc [hunk 26]
```diff
@@ -497,16 +520,32 @@ void Receiver::loopRecv(int tid, int core_id, SampleBuffer* rx_buffer) {
...
+            // if buffer was full, exit
+            if ((old & bit) != bit) {
+              MLPD_ERROR("thread %d buffer full\n", tid);
+              throw std::runtime_error("Thread %d buffer full\n");
+            }
+            // Reserved until marked empty by consumer
```
Comment (asedlmayr) [inline CC/Sounder/receiver.cc:544]: remove comment as it is not accurate

File: CC/Sounder/receiver.cc [hunk 27]
```diff
@@ -497,16 +520,32 @@ void Receiver::loopRecv(int tid, int core_id, SampleBuffer* rx_buffer) {
...
-            !config_->isUlData(frame_id, slot_id))
+            !config_->isUlData(frame_id, slot_id)) {
+          for (size_t ch = 0; ch < num_packets; ++ch) {
+            const int bit = 1 << (cursor + ch) % sizeof(std::atomic_int);
+            //                std::cout<<"bit at 2 is " << bit << std::endl;
+            const int offs = (cursor + ch) / sizeof(std::atomic_int);
+            const int old =
+                std::atomic_fetch_and(&pkt_buf_inuse[offs], ~bit);  // now full
+            //                std::cout<<"old at 2 is " << old << std::endl;
+            // if buffer was full, exit
```
Comment (asedlmayr) [inline CC/Sounder/receiver.cc:539]: Remove bad comment.

File: CC/Sounder/receiver.cc (hunk 26 above)
Comment (asedlmayr) [reply CC/Sounder/receiver.cc:544]: Fix the comments (or use this code chunk)
      for (size_t ch = 0; ch < num_packets; ++ch) {
        const int bit = 1 << (cursor + ch) % sizeof(std::atomic_int);
        const int offs = (cursor + ch) / sizeof(std::atomic_int);
        const int old = std::atomic_fetch_and(&pkt_buf_inuse[offs], ~bit);  // now empty
        // if buffer was empty, exit
        if ((old & bit) != bit) {
          MLPD_ERROR("thread %d freed buffer when already free\n", tid);
          throw std::runtime_error("buffer empty during free");
        }
      }

File: CC/Sounder/receiver.cc [hunk 28]
```diff
@@ -740,12 +787,13 @@ int Receiver::clientTxData(int tid, int frame_id, long long base_time) {
       long long txTime = txFrameTime +
                          config_->cl_ul_slots().at(tid).at(s) * num_samps -
                          config_->tx_advance(tid);
-      if (kUseUHD && s < (tx_slots - 1))
+      if ((kUseUHD || pUseUHD) && s < (tx_slots - 1))
```
Comment (asedlmayr) [inline CC/Sounder/receiver.cc:790]: add brackets to a compound if {}

File: CC/Sounder/receiver.cc [hunk 29]
```diff
@@ -969,6 +1019,7 @@ void Receiver::clientSyncTxRx(int tid, int core_id, SampleBuffer* rx_buffer) {
         // Set buffer status(es) to full; fail if full already
         for (size_t ch = 0; ch < config_->cl_sdr_ch(); ++ch) {
           int bit = 1 << (buffer_offset + ch) % sizeof(std::atomic_int);
+          std::cout << "bit is " << bit << std::endl;
```
Comment (asedlmayr) [inline CC/Sounder/receiver.cc:1022]: remove this debug print

**Review by asedlmayr — COMMENTED** [review]
In general looks pretty good.  Just a few fixes before commit

File: CC/Sounder/include/Radio.h (hunk 16 above)
Comment (Shadowcaster77) [reply CC/Sounder/include/Radio.h:13]: these are still necessary since it relies on some of the flags created by those soapy headers.

File: CC/Sounder/receiver.cc (hunk 26 above)
Comment (Shadowcaster77) [reply CC/Sounder/receiver.cc:544]: thanks for the comment

File: CC/Sounder/include/ClientRadioSetUHD.h (hunk 15 above)
Comment (Shadowcaster77) [reply CC/Sounder/include/ClientRadioSetUHD.h:40]: made corresponding changes

File: CC/Sounder/include/RadioUHD.h (hunk 20 above)
Comment (Shadowcaster77) [reply CC/Sounder/include/RadioUHD.h:14]: still necessary, some "time variable" is still calculated based on the Soapy method

File: CC/Sounder/include/BaseRadioSetUHD.h (hunk 17 above)
Comment (Shadowcaster77) [reply CC/Sounder/include/BaseRadioSetUHD.h:11]: soapy headers are still necessary, some types like SOAPY_SDR_CS16, I still used those.
I checked with deleting the json include, it gives errors.

File: CC/Sounder/ClientRadioSetUHD.cc (hunk 14 above)
Comment (Shadowcaster77) [reply CC/Sounder/ClientRadioSetUHD.cc:17]: That's true, it is kind of not pretty.

File: CC/Sounder/receiver.cc [hunk 30]
```diff
@@ -838,12 +882,12 @@ void Receiver::clientSyncTxRx(int tid, int core_id, SampleBuffer* rx_buffer) {
   int rx_offset = 0;
 
   // For USRP clients skip UHD_INIT_TIME_SEC to avoid late packets
-  if (kUseUHD == true) {
+  if (kUseUHD == true || kUseSoapyUHD == true) {
```
Comment (rdoosty) [inline CC/Sounder/receiver.cc:885]: It would be nice to change `kUseUHD` to `kUsePureUHD`.

File: PYTHON/IrisUtils/plot_hdf5.py [hunk 31]
```diff
@@ -317,6 +315,122 @@ def verify_hdf5(hdf5, frame_i=100, cell_i=0, ofdm_sym_i=0, ant_i =0,
             equalized_symbols, tx_symbols, slot_evm, slot_evm_snr = hdf5_lib.demodulate(ul_samps[:, ul_slot_i, :, :], userCSI, tx_data, metadata, offset, ul_slot_i)
             plot_constellation_stats(slot_evm, slot_evm_snr, equalized_symbols, tx_symbols, ref_frame, ul_slot_i)
 
+        # if demodulate:
```
Comment (rdoosty) [inline PYTHON/IrisUtils/plot_hdf5.py:318]: Can these comments be deleted?

Comment (Shadowcaster77) [reply PYTHON/IrisUtils/plot_hdf5.py:318]: yeah, will delete them.

File: PYTHON/IrisUtils/hdf5_lib.py [hunk 32]
```diff
@@ -896,27 +895,53 @@ def load_tx_data(metadata, dirpath):
...
-        tx_file_names = metadata['TX_FD_DATA_FILENAMES'].astype(str)
+        #
+        # aaa = metadata['TX_FD_DATA_FILENAMES']
+        # print(type(aaa))
+        # aaa = aaa.astype(str)
+        # print(type(aaa))
+        #
+        # manually_file = ['ul_data_f_16QAM_52_64_10_1_1_AB_0.bin']
+        # manually_file = ['ul_data_f_QPSK_52_64_10_1_1_AB_0.bin']
+        manually_file = ['ul_data_f_64QAM_52_64_10_1_1_AB_0.bin']
```
Comment (rdoosty) [inline PYTHON/IrisUtils/hdf5_lib.py:906]: why is this manually set?

Comment (Shadowcaster77) [reply PYTHON/IrisUtils/hdf5_lib.py:906]: cause somehow in the saved hdf5 file, for uplink transmissions, the `TX_FD_DATA_FILENAMES` section is missing. So for the testing purposes, I just set it manually.  An example output can be seen [https://docs.google.com/presentation/d/1AFXKeAOm8HbT-o-kMyGJTGtRnMOKC2cAyBhUI0VPuAs/edit?usp=sharing](url)

Comment (rdoosty) [reply PYTHON/IrisUtils/hdf5_lib.py:906]: `TX_FD_DATA_FILENAMES` always gets recorded as long as there are uplink files. This feature has been added a long time ago and is mainlined in the `master` branch too. Most likely, the dataset you are showing in the slides was collected using a very old version. It could be also a corner case of the configs. Please test the latest develop branch again with a known good json and your own json, and let me know if you see this bad behavior

Having said that, manually setting the file name is obviously bad and should not be done. It also broke my tests. So please revert to the original way which is reading the file from the HDF5 dataset. If there is an issue with the Sounder not recording the filename in the metadata, we should fix that instead.

Comment (Shadowcaster77) [reply PYTHON/IrisUtils/hdf5_lib.py:906]: Thanks for providing the detailed explanation, will test and look through it again

Comment (Shadowcaster77) [reply PYTHON/IrisUtils/hdf5_lib.py]: I figured out where the issue mentioned in the previous slides comes from. 

It comes up when I use `--bs_only` and `--client_only` flags when running sounder; in this case, the BS will have no reference to where the TX_FD_DATA_FILENAMES the client is referring to and will not include the `TX_FD_DATA_FILENAMES`

Comment (Shadowcaster77) [reply PYTHON/IrisUtils/hdf5_lib.py]: I just converted the post-processing python files I changed back to the original develop version.

**Review by asedlmayr — APPROVED** [review]
Looks good
